matscipy/dislocation.py

```diff
@@ -2150,6 +2152,242 @@ def check_duplicates(atoms, distance=0.1):
+        if center is None:
+            center = np.diagonal(bulk.get_cell())/2
+        coordinates = [ vec-center for vec in bulk.get_positions()]
+
+        # Find initial Stroh displacements
+        disp1 = self.stroh_solve(coordinates) 
+        if not self_consistent:
+            return disp1 
+
+        # Find self-consistent displacements
+        res = np.inf
+        i = 0
+        while res > tol:
+            coordinates += disp1
+            disp2 = self.stroh_solve(coordinates)
+            res = np.abs( disp1 - disp2 ).max()
+            if verbose:
+                print('disloc SCF', i, '|d1-d2|_inf =', res)
+            disp1 = disp2.copy()
+            i += 1
+            if i > max_iter:
+                raise RuntimeError('Self-consistency' +
+                                   f'did not converge in {max_iter} cycles')
+
+        return disp1  
+
+
+    def deformation_gradient(self, bulk, center=None):
+        """
+        3D displacement gradient tensor of the dislocation. 
+
+        Parameters
+        ----------
+        bulk : ASE atoms object
+            ASE bulk cell whose axes are aligned with self.axes
+        center : 3x1 array
+            Position of the dislocation core within the cell
+
+        Returns
+        -------
+        du_dx, du_dy, du/dz, dv_dx, dv_dy, dv/dz, dw/dx, dw/dy, dw/dz : 1D arrays
+            Displacement gradients:
+            dx, dy, dz: changes in dislocation core position along the three axes of self.axes
+            du, dv, dw: changes in displacements of atoms along the three axes of self.axes, in response to changes in dislocation core position
+        """
+
+        # Get atomic positions wrt dislocation core in Cartesian coordinates
+        if center is None:
+            center = np.diagonal(bulk.get_cell())/2
+        coordinates = [ vec-center for vec in bulk.get_positions()]
+
+        signs = np.sign(np.imag(self.Np))
+        signs[np.where(signs==0.0)] = 1.0
+        
+        A = self.Nv[0:3,:]
+        L = self.Nv[3:6,:]
+        D = signs * np.einsum('i,ij', self.b, L)
+
+        eta = (np.expand_dims(np.einsum('i,ji', self.m, coordinates), axis=1)
+               + np.outer(np.einsum('i,ji', self.n, coordinates), self.Np))
+        
+        pref = (1.0/(2.0 * np.pi * 1.0j))
+
+        # Compute the displacement gradient components
+        nat = len(coordinates)
+        grad3D_T = np.zeros((3,3,nat))
+        for i in range(3):
+            for j in range(3):
+                grad3D_T[i,j,:] = np.real(( pref * np.einsum('ij,j', 1/eta, (self.m[j] + self.Np * self.n[j]) * A[i,:] * D) ))
+
+        # Add unity matrix along the diagonal block, to turn this into the deformation gradient tensor.
+        for i in range(3):
+            grad3D_T[i,i,:] += np.ones(nat)
+        
+        # Transpose to get the correct shape
+        grad3D = np.transpose(grad3D_T)
+
+        return grad3D
+        
+
+    def deformation_gradient_2D(self, bulk, center=None):
```

Comment:
This could be combined with the method above by adding an optional `return_2D=False` bool or similar, like the `full_outputs=False` used in some numpy/scipy routines.

Reply:
Done, grad2D is now an optional output of the deformation_gradient() function.